docs/DafnyRef/Refinement.md

```diff
@@ -92,32 +92,133 @@ TODO
 
 ## 21.4. Const declarations
 
-A parent `const` declaration may be refined by a refining `const` declaration
-if
+A child `const` declaration may refine a `const` declaration from a
+parent module if
 
 * the parent has no initialization,
 * the child has the same type as the parent, and
 * one or both of the following holds:
    * the child has an initializing expression
-   * the child is declared `ghost` and the parent is not `ghost`, or vice versa
+   * the child is declared `ghost` and the parent is not `ghost`, or vice versa.
 
-To indicate it is a refining declaration, a refining `const` declaration
-contains an ellipsis in this syntax:
-```grammar
-"const" { Attribute } CIdentType "..." [ ":=" Expression ]
-```
+A refining module can also introduce new `const` declarations that do
+not exist in the refinement parent.
 
 ## 21.5. Method declarations
 
-TODO
+A child `method` definition may refine a parent `method` declaration or
+definition to
+
+* provide a body missing in the parent,
+* replace the body of the parent with a new, semantically-compatible
+  body,
```

Comment:
Is this true? Going from memory, I think a body in the parent can only be changed by doing a combination of:

* making the child more deterministic
* changing `assume` statements into `assert` statements
* superimposing new statements onto the parent (you use the word "amend", whereas the literature tends to say "superposition refinement", and I might say "superimpose")

Reply:
!! That's a great example and if it's intended it should be in the manual: my expectation re. refinement is that behaviors of the child are found in the parent, but this isn't the case here.

Reply:
Oh, this is probably just due to the implicit `...` before the closing `}` combined with the rule that additional `return` statements are allowed!

Reply:
Ah, so the method becomes `return x + 1; return x + 2;` after this?

Reply:
My current understanding is that the implicit `...` at the end would match `return x + 1`, so it'd actually be `return x + 2; return x + 1;`

Reply:
Indeed, it prints 2, thanks!

Reply:
I guess the trick is that the visible part of the refinement is a refinement of the spec, not of the body, since method bodies are always opaque.  Hence refining x+1 into x+2 is valid as long as the spec of the `x+1` part is vague enough to also allow x+2.  And this doesn't work for functions (whose body is transparent) because you're not allowed to change function bodies.  Interesting.

Reply:
Correctly deduced. Because of the implicit `...` before the `}`, the method body in `M1` becomes `return x + 1; return x + 2;`. It would be sound to let a method replace its entire body with something different, but there is no special syntax for doing so.